File: queue.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/malloc.h>

/*
定义节点
包括数据data、指向下一节点的pNext指针
*/
typedef struct node
{
	int data;
	struct node *pNext;
}NODE, *PNODE;

/*
定义队列
*/
typedef struct queue
{
	PNODE pFront;  //队头  用于出队
	PNODE pRear;  //队尾  用于入队
}*QUEUE;

/*
队列初始化操作
*/
QUEUE init(){
	QUEUE q;

	//申请一个无实际意义的节点，用于队列的初始化
	PNODE pNew = (PNODE)malloc(sizeof(NODE));
	if (pNew == NULL)
	{
		printf("初始化-----------分配内存失败！\n");
		exit(0);
	}
	pNew->pNext = NULL;
	pNew->data = 0;

	q->pFront = pNew;
	q->pRear = pNew;

	return q;
}

/*
入队操作
*/
void ins_queue(QUEUE q, int data){
	//申请一个节点
	PNODE pNew = (PNODE)malloc(sizeof(NODE));
	if (pNew == NULL)
	{
		printf("入队----------申请新节点失败\n");
	}
	pNew->data = data;
	pNew->pNext = NULL;

	//在队尾入队
	q->pRear->pNext = pNew;
	q->pRear = pNew;
}

/*
出队操作
*/
int out_queue(QUEUE q){
	//从队头出队  front->pNext永远都指向队头的元素
	PNODE pTmp = q->pFront->pNext;
	int data = pTmp->data;
	q->pFront->pNext = pTmp->pNext;  //让pFront的next指针域指向原链表的下一个节点


	free(pTmp);
	return data;
}

/*
遍历操作
*/
void travel(QUEUE q){
	PNODE pTmp = q->pFront->pNext;  //找到指向的第一个节点
	int i = 0;
	while (pTmp != NULL)
	{
		printf("遍历-----------输出第%d个节点，节点数据为：%d\n", i, pTmp->data);
		pTmp = pTmp->pNext;
		i++;
	}
}


int main(int argc, char const *argv[])
{
	QUEUE q = init();
	ins_queue(q, 1);
	ins_queue(q, 2);
	ins_queue(q, 3);
	ins_queue(q, 4);

	printf("-------------开始遍历---------------\n");
	travel(q);
	printf("--------------结束遍历---------------\n\n");

	printf("出队一个元素：\n");
	int o1 = out_queue(q);
	printf("出队元素为：%d\n", o1);
	printf("出队完成。\n\n");

	printf("-------------开始遍历---------------\n");
	travel(q);
	printf("--------------结束遍历---------------\n\n");



	return 0;
}

File: phpCode/queue.php
<?php
/*
 * 节点
 * 包括数据、下一节点
 * */
class Node{
    public $data;
    public $next_node;
    
    public function __construct($data, $node=null){
        $this->data = $data;
        $this->next_node = $node;
    }
    
    public function setNextNode($Node){
        $this->next_node = $Node;
    }
}

/*
 * 队列
 * */
class Queue{
    public $size;  //队列中节点的个数
    public $front;  //队头
    public $rear;  //队尾
    
    /*
     * 入队  在队尾入队
     * */
    public function enQueue($data){
        $node = new Node($data);  //新节点  该节点next节点为空
        if ($this->size == 0){  //空队列 直接将front、rear指向该节点
            $this->rear = $node;
            $this->front = $node;
        } else {
            $this->rear->setNextNode($node);  //将该新节点 挂载到 上一个节点
            $this->rear = $node;  //将队尾指向该新节点
        }
        
        $this->size++;
    }
    
    /*
     * 出队  队头出队
     * */
    public function outQueue(){
        $res = $this->front->data;
        
        //将该队头指向下一个节点
        $this->front = $this->front->next_node;
        $this->size--;
        
        return $res;
    }
    
    public function size(){
        return $this->size;
    }
}

$queue = new Queue();
$queue->enQueue(1);
$queue->enQueue(9);
$queue->enQueue(5);
$queue->enQueue(3);
$queue->enQueue(7);

var_dump($queue->outQueue());
var_dump($queue->size());
var_dump($queue->outQueue());
var_dump($queue->size());





